File: OGRSH/src/jcomm/DefaultOGRSHReadBuffer.cpp
#include <string>

#include "ogrsh/Logging.hpp"

#include "jcomm/IPackable.hpp"
#include "jcomm/IOGRSHReadBuffer.hpp"
#include "jcomm/DefaultOGRSHReadBuffer.hpp"

namespace jcomm
{
	std::string DefaultOGRSHReadBuffer::readUTF()
		throw (IOException)
	{
		int length = _buffer.getInt();
		char *data = new char[length + 1];
		_buffer.get(data, 0, length);
		data[length] = (char)0;
		std::string str(data);
		delete []data;
		return str;
	}

	void DefaultOGRSHReadBuffer::verifyType(
		const std::string &expectedType)
	{
		std::string type = readUTF();
		verifyType(type, expectedType);
	}

	void DefaultOGRSHReadBuffer::verifyType(
		const std::string &type,
		const std::string &expectedType)
	{
		if (type == expectedType)
			return;

		OGRSH_FATAL("Expected type \"" << expectedType
			<< "\" but received type \"" << type << "\".");
		ogrsh::shims::real_exit(1);
	}

	DefaultOGRSHReadBuffer::DefaultOGRSHReadBuffer(ByteBuffer &buffer)
		: _buffer(buffer)
	{
	}

	void DefaultOGRSHReadBuffer::readBoolean(bool &target)
		throw (IOException)
	{
		verifyType("java.lang.Boolean");
		char c = _buffer.get();
		if (c == (char)0)
			target = false;
		else
			target = true;
	}

	void DefaultOGRSHReadBuffer::readChar(char &target)
		throw (IOException)
	{
		verifyType("java.lang.Byte");
		target = _buffer.get();
	}

	void DefaultOGRSHReadBuffer::readShort(short &target)
		throw (IOException)
	{
		verifyType("java.lang.Short");
		target = _buffer.getShort();
	}

	void DefaultOGRSHReadBuffer::readInt(int &target)
		throw (IOException)
	{
		verifyType("java.lang.Integer");
		target = _buffer.getInt();
	}

	void DefaultOGRSHReadBuffer::readLongLong(long long &target)
		throw (IOException)
	{
		verifyType("java.lang.Long");
		target = _buffer.getLongLong();
	}

	void DefaultOGRSHReadBuffer::readFloat(float &target)
		throw (IOException)
	{
		verifyType("java.lang.Float");
		target = _buffer.getFloat();
	}

	void DefaultOGRSHReadBuffer::readDouble(double &target)
		throw (IOException)
	{
		verifyType("java.lang.Double");
		target = _buffer.getDouble();
	}

	bool DefaultOGRSHReadBuffer::readString(std::string  &target)
		throw (IOException)
	{
		std::string type = readUTF();
		if (type == "null")
			return false;

		verifyType(type, "java.lang.String");
		target = readUTF();
		return true;
	}

	bool DefaultOGRSHReadBuffer::readPackable(IPackable &target)
		throw (IOException)
	{
		std::string type = readUTF();
		if (type == "null")
			return false;

		verifyType(type, target.getTypeName());
		target.unpack(*this);
		return true;
	}

	void DefaultOGRSHReadBuffer::readBytes(void **buf, ssize_t &length)
		throw (IOException)
	{
		std::string type = readUTF();
		if (type == "null")
		{
			length = 0;
			return;
		}

		verifyType(type, "byte-array");
		length = _buffer.getInt();
		_buffer.get(*((char**)buf), 0, length);
	}
}

File: src/edu/virginia/vcgr/ogrsh/server/packing/DefaultOGRSHReadBuffer.java
package edu.virginia.vcgr.ogrsh.server.packing;

import java.io.IOException;
import java.lang.reflect.Array;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.nio.ByteBuffer;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

public class DefaultOGRSHReadBuffer implements IOGRSHReadBuffer
{
	static private Log _logger = LogFactory.getLog(DefaultOGRSHReadBuffer.class);
	
	private ByteBuffer _source;

	private Class<?> findClass(String className) throws IOException
	{
		try
		{
			return Thread.currentThread().getContextClassLoader().loadClass(className);
		}
		catch (ClassNotFoundException cnfe)
		{
			_logger.error("Can't deserialize type \"" + className + "\".", cnfe);
			throw new IOException("Can't deserialize type \"" + className 
				+ "\".");
		}
	}
	
	private IPackable construct(Class<? extends IPackable> type) throws IOException
	{
		Constructor<? extends IPackable> cons = null;
		
		try
		{
			try
			{
				cons = type.getConstructor(new Class[] { IOGRSHReadBuffer.class });
				return cons.newInstance(new Object[] { this });
			} catch (NoSuchMethodException nsme)
			{
				cons = type.getConstructor(new Class[0]);
				IPackable packable = cons.newInstance(new Object[0]);
				packable.unpack(this);
				return packable;
			}
		}
		catch (NoSuchMethodException nsme)
		{
			_logger.error("Unable to create instance of packable type \""
				+ type.getName() + "\".", nsme);
			throw new IOException(
				"Unable to create instance of packable type \"" 
				+ type.getName() + "\".");
		}
		catch (IllegalAccessException iae)
		{
			_logger.error("Unable to create instance of packable type \""
				+ type.getName() + "\".", iae);
			throw new IOException(
				"Unable to create instance of packable type \"" 
				+ type.getName() + "\".");
		}
		catch (InstantiationException ia)
		{
			_logger.error("Unable to create instance of packable type \""
				+ type.getName() + "\".", ia);
			throw new IOException(
				"Unable to create instance of packable type \"" 
				+ type.getName() + "\".");
		}
		catch (InvocationTargetException ite)
		{
			_logger.error("Unable to create instance of packable type \""
				+ type.getName() + "\".", ite);
			throw new IOException(
				"Unable to create instance of packable type \"" 
				+ type.getName() + "\".");
		}
	}
	
	public DefaultOGRSHReadBuffer(ByteBuffer source)
	{
		_source = source;
	}
	
	public String readUTF() throws IOException
	{
		int length = _source.getInt();
		byte []data = new byte[length];
		_source.get(data);
		return new String(data, "UTF-8");
	}

	@SuppressWarnings("unchecked")
	public Object readObject() throws IOException
	{
		String label = readUTF();
		if (label.equals("null"))
			return null;
		if (label.equals("byte-array"))
		{
			int len = _source.getInt();
			byte []data = new byte[len];
			_source.get(data);
			return data;
		}
		if (label.equals("array"))
		{
			int len = _source.getInt();
			label = readUTF();
			Class<?> componentType = findClass(label);
			Object []ret = Object[].class.cast(Array.newInstance(componentType, len));
			for (int lcv = 0; lcv < ret.length; lcv++)
			{
				ret[lcv] = readObject();
			}
			return ret;
		}
		
		Class<?> type = findClass(label);
		if (type.equals(Boolean.class))
		{
			return new Boolean(_source.get() != (byte)0);
		} else if (type.equals(Byte.class))
		{
			return new Byte(_source.get());
		} else if (type.equals(Character.class))
		{
			return new Character(_source.getChar());
		} else if (type.equals(Short.class))
		{
			return new Short(_source.getShort());
		} else if (type.equals(Integer.class))
		{
			return new Integer(_source.getInt());
		} else if (type.equals(Long.class))
		{
			return new Long(_source.getLong());
		} else if (type.equals(Float.class))
		{
			return new Float(_source.getFloat());
		} else if (type.equals(Double.class))
		{
			return new Double(_source.getDouble());
		} else if (type.equals(String.class))
		{
			return readUTF();
		} else if (IPackable.class.isAssignableFrom(type))
		{
			return construct((Class<? extends IPackable>)type);
		} else
		{
			_logger.error("Don't know how to deserialize type \"" 
				+ type.getName() + "\".");
			throw new IOException("Don't know how to deserialize type \"" 
				+ type.getName() + "\".");
		}
	}
}